Import Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

Create path

In [2]:
path = r'/Users/chrisherin/Documents/01_Jupyter/11-2021 Instacart Basket Analysis'

In [3]:
path

'/Users/chrisherin/Documents/01_Jupyter/11-2021 Instacart Basket Analysis'

Import data

:: Orders dataframe

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

:: Products dataframe

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

:: import check

In [52]:
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [7]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

:: drop irrelevant columns from df_ords

In [42]:
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  orders_day_of_week  \
0        2539329        1             1                   2   
1        2398795        1             2                   3   
2         473747        1             3                   3   
3        2254736        1             4                   4   
4         431534        1             5                   4   
...          ...      ...           ...                 ...   
3421078  2266710   206209            10                   5   
3421079  1854736   206209            11                   4   
3421080   626363   206209            12                   1   
3421081  2977660   206209            13                   1   
3421082   272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

:: save df without column 'eval_set'

In [44]:
df_ords = df_ords.drop(columns = ['eval_set'])

In [45]:
# check new df_ords
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

How to check values in a column, find NA values.   df['variable'].value_counts(dropna = False)

In [46]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

Renaming Columns

In [55]:
# {'existing_name' : 'new_name'}, (inplace = True) is the operator meaning overwrite the existing name with new name
df_ords.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [56]:
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

Changing a variables data type

In [12]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [13]:
df_ords['order_id'].dtype

dtype('O')

Transposing Data

:: Import 'Departments' data

In [14]:
path = r'/Users/chrisherin/Documents/01_Jupyter/11-2021 Instacart Basket Analysis'

In [15]:
path

'/Users/chrisherin/Documents/01_Jupyter/11-2021 Instacart Basket Analysis'

In [16]:
df_dep = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [17]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

:: Transpose from wide to deep, moving data from 1 row/22 colums - 22 rows/2 columns

In [18]:
# .T = opperator to transpose data from 'wide' to 'deep'
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

:: Overwrite df_dep for use as new transposed df_dep_t

In [19]:
df_dep_t = df_dep.T

In [20]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

Resetting index. 
Header on new transposed df shows '0', need to remove and index the rows.

:: First index the rows

In [21]:
# This will index the rows of the transposed data, using .reset_index()
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

:: Step 1, new header

In [22]:
# .iloc[] = LOCate an Integer. Using this function, you can locate the row to assign a new header. 
# Note, brakets not parenthesis
new_header = df_dep_t.iloc[0]

In [23]:
new_header

0    department
Name: department_id, dtype: object

:: Step 2, new df. Copy the existing df from row 1 on to end

In [24]:
# the function [1:] tells python to copy from row 1 to end
df_dep_t_new = df_dep_t[1:]

In [25]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

:: Step 3, insert new_header

In [26]:
# using .columns you can insert new names to columns. 
# In this line, by using 'new_header' that you created, columns are renamed to the new header
df_dep_t_new.columns = new_header

In [27]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

Note: the dataframe df_dep_t_new is missing the index column here. This could be added by .reset_index() . 

Create Data Dictionary

In [28]:
# use function .to_dict('index') to turn df into dictionary. Make sure to assign it to a variable such as 'data_dict'
data_dict = df_dep_t_new.to_dict('index')

In [29]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

The to_dict() function transformed your df_dep_t_new dataframe into dictionary format and saved it in a new variable, data_dict. The argument, index, tells Python to use the numbered rows as the key values for the entries in the dictionary.

:: Test and check

In [30]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

Q: what department is Chocolate Sandwich Cookies in?

In [31]:
print(data_dict.get('19'))

{'department': 'snacks'}


Chocolate Sandwich Cookies are in the Snacks department!

Subsetting

:: Indexing - to retrieve an element from an object. syntax: object['element']

In [32]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

:: Nesting syntax above inside df[ ] will return rows where only the [dept_id] == 19 is True

In [33]:
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

:: Once the filter is defined you can create the subset df - df_snacks = df[df['element']==identifier]

In [34]:
df_snacks = df_prods[df_prods['department_id']==19]

:: Dataframe = all items in the snack department!

In [36]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

:: Same subsetting using loc function. loc will look for the column. 

In [38]:
# Creating same df with different syntax and name for practice
df_snacks_2 = df_prods.loc[df_prods['department_id']==19]

In [39]:
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

:: Another alt for subsetting a df - using .isin can pull multiple identifiers

:: Ex - .isin[(17,18,19)]

In [40]:
# Creating same df with different syntax and name for practice
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [41]:
df_snacks_3.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

Exporting dataframes

:: syntax ex - df_ords.to_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'))

::

::

Task Work

2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

:: user_id should be changed to str

In [48]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [49]:
df_ords['user_id'].dtype

dtype('O')

3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

In [57]:
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

:: most column names are pretty intuitive but for the sake of the exercise I can change days_since_prior_order to days_since_last_order as another column.

In [58]:
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = False)

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
1        2398795       1             2                   3                  7   
2         473747       1             3                   3                 12   
3        2254736       1             4                   4                  7   
4         431534       1             5                   4                 15   
...          ...     ...           ...                 ...                ...   
3421078  2266710  206209            10                   5                 18   
3421079  1854736  206209            11                   4                 10   
3421080   626363  206209            12                   1                 12   
3421081  2977660  206209            13                   1                 12   
3421082   272231  206209            14                   6                 14   

         days_since_last_order  
0                          NaN  
1                         15.0  
2                         21.0  
3                         29.0  
4                         28.0  
...                        ...  
3421078                   29.0  
3421079                   30.0  
3421080                   18.0  
3421081                    7.0  
3421082                   30.0  

[3421083 rows x 6 columns]

:: Unsure how/why it overwrote the column instead of created a copy...

4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [90]:
df_ords.mode()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0              1  100142           1.0                 0.0               10.0   
1             10  100316           2.0                 NaN                NaN   
2            100  100325           3.0                 NaN                NaN   
3           1000   10046           4.0                 NaN                NaN   
4          10000  100514           NaN                 NaN                NaN   
...          ...     ...           ...                 ...                ...   
3421078   999995     NaN           NaN                 NaN                NaN   
3421079   999996     NaN           NaN                 NaN                NaN   
3421080   999997     NaN           NaN                 NaN                NaN   
3421081   999998     NaN           NaN                 NaN                NaN   
3421082   999999     NaN           NaN                 NaN                NaN   

         days_since_prior_order  
0                          30.0  
1                           NaN  
2                           NaN  
3                           NaN  
4                           NaN  
...                         ...  
3421078                     NaN  
3421079                     NaN  
3421080                     NaN  
3421081                     NaN  
3421082                     NaN  

[3421083 rows x 6 columns]

:: the time of day when the most orders come in is 10am

Note: unsure if this is correct, or how to limit the function to just 'order_hour_of_day'

5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [92]:
print(data_dict.get('4'))

{'department': 'produce'}


6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [94]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [95]:
df_breakfast = df_prods[df_prods['department_id']==14]

In [97]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

7. They’d also like to see details about customers who might be throwing dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client

In [98]:
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [100]:
df_dinner_parties.head(10)

product_id                                       product_name  aisle_id  \
2            3               Robust Golden Unsweetened Oolong Tea        94   
6            7                     Pure Coconut Water With Orange        98   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
16          17                                  Rendered Duck Fat        35   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1  
19              7     6.0  
22             12     8.2  
34             12    15.1  
38              7    12.5  
39             12    22.5

8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [102]:
df_dinner_parties.shape

(7650, 5)

:: df_dinner_parties contains 7650 rows

9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [113]:
df_user1 = df_ords[df_ords['user_id']=='1']

In [114]:
df_user1.head(20)

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

:: forgot that the type for user_id was changed to str so the number needed to be put in ''

:: created a subset for further analysis

10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [116]:
df_user1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

:: all stats for user_1 

In [117]:
df_ords.to_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'))

In [118]:
df_dep_t_new.to_csv(os.path.join(path,'02 Data','Prepared Data','departments_wrangled.csv'))

:: Task complete